### ISSUE TITLE
Support PCA in ANNlite

### ISSUE THREAD

**Issue by jemmyshin** [jina-ai/annlite#134]
In order to reduce memory usage, we need to implement PCA inside `ANNlite`. There will be two PRs for this function:

1. implement PCA based on `scikit-learn`
2. integrate PCA with `ANNlite`

**Comment by JoanFM** [comment]
How will this be integrated? is it a way to do dimensionality reduction? How will u mantain `ANNLite` being incremental?

**Comment by numb3r3** [comment]
pca will be integrated as a feature provided by annlite. It provides the function to reduce the embedding dimension so as the memory footprint. At this moment, we make a strong assumption that the training data (for pca) comes from the same distribution of indexed data, and the incremental data distribution would not have a large shift.

**Comment by JoanFM** [comment]
But why is this better than PQ? We have PQ already for "dim reduction" right?

**Comment by numb3r3** [comment]
Generally, we want to take advantage of the fact that PCA dimensionality reduction preserves more information in the first components. PCA theoretically can achieve better ANN results with the same compress rate. In practice, pca and pq are usually used together (pca first, and then pq) to achieves good accuracy–space trade-offs. And we also find research work that demonstrates the combination idea works https://cs.uwaterloo.ca/~jimmylin/publications/Ma_etal_EMNLP2021.pdf

**Comment by tommykoctur** [comment]
Hello, 

would the PCA work with multiple shards of ANNLITE ? And how ? Will there will be a different projections for each shard. Would it be ok ?

Thanks

**Comment by numb3r3** [comment]
A good practice is to prepare a PCA ahead (offline) which can be shared across shards.

**Comment by tommykoctur** [comment]
Hi @numb3r3, I briefly check the Projector code and it doesn't seems to me that there is some prepare loading of pretrained PCA. Or am I missing something ?

**Comment by numb3r3** [comment]
@jemmyshin do we have such document or examples?
